tower-http/src/sensitive_header.rs

```diff
@@ -0,0 +1,224 @@
+//! Middlewares that mark headers as [sensitive].
+//!
+//! [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+
+use futures_util::ready;
+use http::{header::HeaderName, Request, Response};
+use pin_project::pin_project;
+use std::future::Future;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Mark a header as [sensitive] on both requests and responses.
+///
+/// [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+#[derive(Clone, Debug)]
+pub struct SetSensitiveHeaderLayer {
+    header: HeaderName,
+}
+
+impl SetSensitiveHeaderLayer {
+    /// Create a new [`SetSensitiveHeaderLayer`].
+    pub fn new(header: HeaderName) -> Self {
+        Self { header }
+    }
+}
+
+impl<S> Layer<S> for SetSensitiveHeaderLayer {
+    type Service = SetSensitiveRequestHeader<SetSensitiveResponseHeader<S>>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetSensitiveRequestHeader::new(
+            SetSensitiveResponseHeader::new(inner, self.header.clone()),
+            self.header.clone(),
+        )
+    }
+}
+
+/// Mark a request header as [sensitive].
+///
+/// [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+#[derive(Clone, Debug)]
+pub struct SetSensitiveRequestHeaderLayer {
+    header: HeaderName,
+}
+
+impl SetSensitiveRequestHeaderLayer {
+    /// Create a new [`SetSensitiveRequestHeaderLayer`].
+    pub fn new(header: HeaderName) -> Self {
+        Self { header }
+    }
+}
+
+impl<S> Layer<S> for SetSensitiveRequestHeaderLayer {
+    type Service = SetSensitiveRequestHeader<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetSensitiveRequestHeader {
+            inner,
+            header: self.header.clone(),
+        }
+    }
+}
+
+/// Mark a request header as [sensitive].
+///
+/// [sensitive]: https://docs.rs/http/latest/http/header/struct.HeaderValue.html#method.set_sensitive
+#[derive(Clone, Debug)]
+pub struct SetSensitiveRequestHeader<S> {
+    inner: S,
+    header: HeaderName,
+}
+
+impl<S> SetSensitiveRequestHeader<S> {
+    /// Create a new [`SetSensitiveRequestHeader`].
+    pub fn new(inner: S, header: HeaderName) -> Self {
+        Self { inner, header }
+    }
+
+    /// Gets a reference to the underlying service.
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying service.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes `self`, returning the underlying service.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
```

Comment:
Fun fact. I actually had such a macro way back but I decided to remove it since macros aren't everyones cup of tea. But I think tidying this up with a macro would be cool. I'm gonna make a mental note of that.